ap/calibration/jets.py

```diff
@@ -0,0 +1,432 @@
+
+    # pass text files to calibrator method
+    for key in ("jec", "junc"):
+        # pass the paths to the text files that contain the corrections/uncertainties
+        files = call_kwargs[f"{key}_files"] = [
+            t.path for t in resolve_sample(bundle.files[key])
+        ]
+
+        # also pass a list of tuples encoding the correspondence between the original
+        # file basenames and filenames on disk (as determined by `BundleExternalFiles`)
+        # and the original file basenames (needed by coffea tools to handle the
+        # weights correctly)
+        basenames = get_basenames(files)
+        orig_basenames = get_basenames(resolve_sample(task.config_inst.x.external_files[key]))
+        call_kwargs[f"{key}_names"] = list(zip(basenames, orig_basenames))
+
+    # ensure exactly one 'UncertaintySources' file is passed
+    if len(call_kwargs["junc_names"]) != 1:
+        raise ValueError(
+            f"Expected exactly one 'UncertaintySources' file, got {len(call_kwargs['junc_names'])}",
+        )
+
+    # update the weight names to include the uncertainty sources specified in the config
+    call_kwargs["junc_names"] = [
+        (f"{basename}_{src}", f"{orig_basename}_{src}")
+        for basename, orig_basename in call_kwargs["junc_names"]
+        for src in task.config_inst.x.jec["uncertainty_sources"]
+    ]
+
+
+#
+# Jet energy resolution smearing
+#
+
+@calibrator(
+    uses={
+        "nJet", "Jet.pt", "Jet.eta", "Jet.phi", "Jet.mass", "Jet.genJetIdx", "fixedGridRhoFastjetAll",
+        "nGenJet", "GenJet.pt", "GenJet.eta", "GenJet.phi",
+    },
+    produces={
+        "Jet.pt", "Jet.mass",
+        "Jet.pt_unsmeared", "Jet.mass_unsmeared",
+    } | {
+        f"Jet.{shifted_var}_JERSF_{jersf_dir}"
+        for shifted_var in ('pt', 'mass')
+        for jersf_dir in ('up', 'down')
+
+    },
+)
+def jer(events, config_inst, dataset_inst, jer_files, jersf_files, jer_names, jersf_names, **kwargs):
+    """Apply jet energy resolution smearing and calculate shifts for JER scale factor variations.
+    Follows the recommendations given in https://twiki.cern.ch/twiki/bin/viewauth/CMS/JetResolution."""
+
+    # save the unsmeared properties in case they are needed later
+    set_ak_column(events, "Jet.pt_unsmeared", events.Jet.pt)
+    set_ak_column(events, "Jet.mass_unsmeared", events.Jet.mass)
+
+    # use event numbers in chunk to seed random number generator
+    rand_gen = np.random.Generator(np.random.SFC64(events.event.to_list()))
+
+    # build/retrieve lookup providers for JECs and uncertainties
+    jer_provider = get_lookup_provider(
+        jer_files,
+        coffea_txt_converters.convert_jr_txt_file,
+        coffea_jetmet_tools.JetResolution,
+        names=jer_names,
+    )
+    jersf_provider = get_lookup_provider(
+        jersf_files,
+        coffea_txt_converters.convert_jersf_txt_file,
+        coffea_jetmet_tools.JetResolutionScaleFactor,
+        names=jersf_names,
+    )
```

Comment:
Should both `jer_provider` and `jersf_provider` be created once in the setup function? Or maybe I don't see the dependence in event content.

Reply:
You're right, there is no dependence on the event content here and this can definitely be moved to `setup`. (One might have to revert this if we ever get JECs again that don't align nicely with the eras and we need to check the event content to get the run, but then some other things would also need to change too.)

Reply:
Ah right.... run-dependence of data comes to mind. Since the calls are memoized are thus hopefully fast enough (esp. compared to the actual JEC runtime), maybe it's even better to leave them here :)

Reply:
OK, no harm leaving them here for the time being. I've added an explanatory comment in case we do decide to refactor this in the future.